README.md
Introduction
============

These scripts were developed as part of an un-confrence session at the Analytics Camp held in
Chapel Hill, NC on February 25, 2012, described on this page as [http://wiki.analyticscamp.org/nc2011proposals](http://wiki.analyticscamp.org/nc2011proposals).

The scripts demonstrate how to pull data from the Google Analytics API to generate an RSS feed
of a website or blog's most popular pages, as well as creating a CSV file for further processing.

* During the presentation, the RSS feed was incorporated into a WordPress blog by using the RSS widget built into the theme I was using.
* Non-programmer analytics geeks also appreciated having their data in a spreadsheet so they could
do their magic.
* Bear in mind, the data these tools collect are only as good as your Google Analytics implementation.

The presentation is available at Slideshare at: http://bitly.com/gaapi2mpp?git

## Installation ##
This example code was provided in both Python and Perl.

* The Python example script is stand-alone and does NOT need the Perl example script nor its dependencies.
* The Perl example script is stand-alone and does NOT need the Python example script nor its dependencies.

Below are the installation instructions for each respectively:

### Python ###

The Python script requires the following libraries:

* [python-googleanalytics](https://github.com/clintecker/python-googleanalytics)
* [PyRSS2Gen](http://www.dalkescientific.com/Python/PyRSS2Gen.html)

You should be able to type in the following after checking-out the source:

Theoretically you should be able to type the following after checking out the source:

`sudo python setup.py install`

You can also just install using `easy_install` like so:

`sudo easy_install python-googleanalytics`
`sudo easy_install PyRSS2Gen`


### Perl ###

The Python script requires the following CPAN libraries:

* [Net::Google::Analytics](http://search.cpan.org/dist/Net-Google-Analytics/)
* [Net::Google::AuthSub](http://search.cpan.org/~simonw/Net-Google-AuthSub-0.5/lib/Net/Google/AuthSub.pm)
* [XML::FeedPP](http://search.cpan.org/~kawasaki/XML-FeedPP-0.43/lib/XML/FeedPP.pm)

Installation is pretty straight forward:

`perl -MCPAN -e shell install Net::Google::Analytics Net::Google::AuthSub XML::FeedPP`

NOTE: 
You may encounter an error installing Net:Google::AuthSub in the form of a file 
permission error generated by ExtUtil::MakeMaker with regards to man page paths.

You'll want to see this page: 

* http://search.cpan.org/~leont/Module-Build-0.40/lib/Module/Build.pm
	
It explains how you can modify your ~/.cpan/CPAN/MyConfig.pm to modify makepl_arg arg.

Here's what mine looked like:

  `'makepl_arg' => q[PREFIX=~/perlmods INSTALLMAN1DIR=~/perlmods/share/man/man1 INSTALLSITEMAN1DIR=~/perlmods/share/man/man1 INSTALLMAN3DIR=~/perlmods/share/man/man3 INSTALLS
ITEMAN3DIR=~/perlmods/share/man/man3],`

On a personal note, Dreamhost doesn't seem to currently want to incorporate this directive from the MyConfig.pm file.


## Operation ##

Regardless of whether you're using the Perl or Python example scripts, you'll want to familiarize 
yourself with the arguments available on the Googl API 
[Dimensions & Metrics Reference](http://code.google.com/apis/analytics/docs/gdata/dimsmets/dimsmets.html) page.

Also note that regardless of whether you run the Python or Perl example script, you'll need
to have access to a Google Analytics account, and have obtained a profile id 
(note, this is NOT the same thing as the ua-key).

The general syntax for both scritps is:
[script name] [gmail/username] [password] [ga profile id]

As discussed in the demonstration, the script is made so it can run regularly using crontab. 
The online '[visual crontab utility](http://www.corntab.com/pages/crontab-gui)' may provide assistance to 
thos individuals who can't always remember the cron syntax.

### Ruby ###

Below is the command-line syntax for running the Python example script:
<pre>
python ga-api2mpp.py username\@gmail.com password 1234567
</pre>

The script generates the follwing files:
* mpp-py.rss - a python generated RSS 2.0 compliant file of the top 20 most popular pages
* mpp-py.csv - a python generated tab-delimited file of all pages pulled from Google Analytics

### Perl ###

Below is the command-line syntax for running the Perl example script:
<pre>
perl ga-api2mpp.pl username\@gmail.com password 1234567
</pre>

The script generates the follwing files:

* mpp-pl.rss - a perl generated RSS 2.0 compliant file of the top 20 most popular pages
* mpp-pl.csv - a perl generated tab-delimited file of all pages pulled from Google Analytics


### Environments ###

Once the required libraries/modules are installed, I was able to run both script on the following platforms:

* Mac OS X 10.7
* Ubuntu 11.00
* CentOS 5
* Windows 7

Of course, your mileage may vary!



CHANGES
February 29: 1.0 Released

  * Basic Scripts released based on the following presentation made at Analytics Camp, Chapel Hill NC, Feb 25: 
     http://slidesha.re/gaapi2mpp

February 24: Started Development

ga-api2mpp.py
##############################################################################
# $Id: 
##############################################################################
# ga-api2mpp.py - this script leverages the Google Analytics API
#		http://code.google.com/apis/analytics/docs/gdata/home.html
#
#		it generates 2 files:
#			mpp.rss - an rss of the top 20 stories from your website/blog
#			mpp.csv - a comma delmited file of all the stories from your website/blog 
#
# ASSUMPTIONS:
#		you have access to a Google Analytics (GA) account
#		you've obtained the profile id (not UA key) from the GA dashboard
#		your article paths include /yyyy/mm/dd ...
#		that you've read the following webpage 1st!
#			https://github.com/clintecker/python-googleanalytics/blob/master/USAGE.md
#
# DEPENDENCIES:
# 		python-googleanalytics- https://github.com/clintecker/python-googleanalytics
# 		PyRSS2Gen- http://www.dalkescientific.com/Python/PyRSS2Gen.html
#
# SYNTAX:
#       command  : python ga-api2mpp.py username\@gmail.com password 1234567
#       cronatab : * */12 * * * python ga-api2mpp.py username\@gmail.com password 1234567
#
# TO DO:
#       see if I can't bring in the profile id(s) from the API
#       use a yaml config file to drive the optional parameters
#		of course I need to not hard-code the dates
#		more robust commandline (getops) handling
#		employ utm tracking for outbound links
#			-- such as ?utm_source=ga-app2mpi&utm_medium=rss-widget&utm_campaign=most-popular-pages
#		once you empoy utm tracking, you'll want to employ utm filtering
#		deal with missing atom:link for validation
#			see: http://validator.w3.org/feed/docs/warning/MissingAtomSelfLink.html
#		more robust error & exception handling 
#			-- 5 retries on a get_data, duration of each doubles until all fail
# 		see if we can do anything more optimally w/libs: 
#			googleanalytics.data import DataPoint, DataSet
#			... espcially .aggregates for metrics (which I couldn't get working)
#
# LICENSE:
# 		ga-api2mpp.pl by Dean Peters is licensed under a 
#		Creative Commons Attribution-NonCommercial-ShareAlike 3.0 
#		United States License. For more details, read here:
#		http://creativecommons.org/licenses/by-nc-sa/3.0/us/
#
# WARANTEE:
#		absolutely none. use at your own peril
#
# KVETCH:
#		got a comment, complaint, criticism or contribution?
#		http://twitter.com/deanpeters
#		http://healyourchurchwebsite.com
#
##############################################################################
from googleanalytics import Connection		# google analytics api library for python
import PyRSS2Gen							# create our RSS
import pickle								# debugging
import datetime								# date time formatting
import sys									# manage command line args, also print to stdout (optionally)
import re									# regular expresions

# first, we need to set things up so we can encapsulate descriptions & titles in CDATA blocks
# 	big thanks to this page: http://stackoverflow.com/questions/5371704/python-generated-rss-outputting-raw-html
#   future for atom ref: http://stackoverflow.com/questions/2360641/supporting-pubsubhubbub-with-pyrss2gen
class NoOutput:
    def __init__(self):
        pass
    def publish(self, handler):
        pass

class IPhoneRSS2(PyRSS2Gen.RSSItem):
    def __init__(self, **kwargs):
        PyRSS2Gen.RSSItem.__init__(self, **kwargs)

    def publish(self, handler):
        self.do_not_autooutput_description = self.description
        self.description = NoOutput() # This disables the Py2GenRSS "Automatic" output of the description, which would be escaped.
        self.do_not_autooutput_title = self.title
        self.title = NoOutput() # This disables the Py2GenRSS "Automatic" output of the title, which would be escaped.
        PyRSS2Gen.RSSItem.publish(self, handler)

    def publish_extensions(self, handler):
        handler._out.write('<%s><![CDATA[%s]]></%s>' % ("description", self.do_not_autooutput_description, "description"))
        handler._out.write('<%s><![CDATA[%s]]></%s>' % ("title", self.do_not_autooutput_title, "title"))

# get the comman line args
# SYNTAX is: python ga_api2mpp.py myusername\@gmail.com mypassssword 1234567
username = sys.argv[1]
password = sys.argv[2]
profileid = sys.argv[3]
domain = 'http://healyourchurchwebsite.com/'		# we'll need this for full urls & guids
connection = Connection(username, password)     # pass in id & pw as strings **if** not in config file

# let's go ahead and create the request data variables
# see the following for args: http://code.google.com/apis/analytics/docs/gdata/dimsmets/dimsmets.html
maxresults = 50
account = connection.get_account(profileid)
start_date = datetime.date(2010, 02, 01)
end_date = datetime.date(2011, 02, 28)
metrics = ['uniquePageviews','pageviews','timeOnPage','bounces','entrances','exits']
dimensions = ['pageTitle','pagePath']
sorts = ['-uniquePageviews','-pageviews',]
filters = [
			['uniquePageviews', '>=', '50',]
		]				# yeah, note this one is different

# now let's plug in the variables into the get_data() call
data = account.get_data(
	start_date, 
	end_date, 
	metrics=metrics, 
	dimensions=dimensions, 
	sort=sorts,
	filters=filters)				# n	max_results = maxresults)

# we're going to use this array to help us print a CSV file
csvItems = ["item#\tpageTotalViews\tpageUniqeViews\tpageTimeOnPage\tpageBounces\tpageTitle\tpagePath",]			

# with each row, we'll append a stash of 'items'
rssItems = []			
ii = 0					# our indice counter
# now let's walk through all the data we collected
for row in data.list:

	# let's convert our datapoints into 2 dictionaries we can de-reerence
	dic_dimensions	=	dict(zip(dimensions, row[0]))
	dic_metrics		=	dict(zip(metrics, row[1]))
	
	# let's dereference the data into string variables, we're going to need this

	# NOTE *** this assumes your article paths include /yyyy/mm/dd 
	pagePath = dic_dimensions['pagePath']
	s_obj = re.search(r"^\/(20[0-1]\d)\/(\d{2})\/(\d{2})\/", pagePath)	# in perl it's {^/(20[0-1]\d)/(\d{2})/(\d{2})/}
	if s_obj:
		pagePubDate = datetime.datetime(int(s_obj.group(1)), int(s_obj.group(2)), int(s_obj.group(3)), 11, 30)
		ii = ii + 1							# we only count those item's that pass the post-pull testing
	else:
		#	pagePubDate = datetime.datetime.utcnow()					# default/fallback value 
		continue

	# now let's do the rest
	pageTitle = dic_dimensions['pageTitle']
	pageViews = dic_metrics['pageviews'] 
	pageUniqueViews = dic_metrics['uniquePageviews'] 
	pageTimeOnPage = dic_metrics['timeOnPage']
	pageBounces = dic_metrics['bounces']
	 	
	pageDescription = "%s %d - %s %s %s %s %s" % ("Item #", ii, "the article", pageTitle, "was viewed", pageViews, "times.")

	# let's keep the RSS file smaller than the CSV file
	if (ii <= 20):												
		# now let's stuff our array full of item goodness
		rssItems.append(
			IPhoneRSS2(
				title = pageTitle.encode('utf-8'), 
				link = domain + pagePath, 
				description = pageDescription.encode('utf-8'), 		# trust me, you'll want to do this
				guid = PyRSS2Gen.Guid(domain + pagePath + "#" + str(ii)), 
				pubDate = pagePubDate))
 
 	csvItems.append(
 		"%d\t%s\t%s\t%s\t%s\t%s\t%s" % (ii, pageViews, pageUniqueViews, pageTimeOnPage, pageBounces, pageTitle.encode('utf-8'), pagePath))
 	
# this is where we 'stuff' the rss object full of the array of items, plus otehr channel data
# this page was a big help figuring thsi out -> http://pastebin.com/J298tQD9
rss = PyRSS2Gen.RSS2(
   title = 'Most Popular Pages - HYCW',
   link = 'http://www.healyourchurchwebsite.com/',
   description = 'The most popular pages from HealYourChurchWebsite',
   lastBuildDate = datetime.datetime.utcnow(),			# or .now() depending on what you want
   items = rssItems)
   
# push it all out to a file ... then standard i/o
# also a big help on output: http://nullege.com/codes/search/PyRSS2Gen.RSS2.write_xml
rss.write_xml(open("mpp-py.rss", "w"), "utf-8")

# uncomment below if you want to display the rss to the console (stdout)
# rss.write_xml(sys.stdout)

# write out the CSV file
# this page was a helpful - http://stackoverflow.com/questions/899103/python-write-a-list-to-a-file
with open('mpp-py.csv', 'w') as file:
    for row in csvItems:
        file.write("{}\n".format(row))
print pickle.dumps(csvItems)

# Yipeee! We're Done -- that's all folks!
exit(0);


ga-api2mpp.pl
#!/usr/bin/perl
##############################################################################
# $Id: 
##############################################################################
# ga-api2mpp.pl - this script leverages the Google Analytics API
#		http://code.google.com/apis/analytics/docs/gdata/home.html
#
#		it generates 2 files:
#			mpp.rss - an rss of the top 20 stories from your website/blog
#			mpp.csv - a comma delmited file of all the stories from your website/blog
#
# ASSUMPTIONS:
#		you have access to a Google Analytics (GA) account
#		you've obtained the profile id (not UA key) from the GA dashboard
#		your article paths include /yyyy/mm/dd ..
#
#
# DEPENDENCIES:
# 		Net::Google::Analytics - http://search.cpan.org/dist/Net-Google-Analytics/
# 		Net::Google::AuthSub - http://search.cpan.org/~simonw/Net-Google-AuthSub-0.5/lib/Net/Google/AuthSub.pm
#       XML::FeedPP	- http://search.cpan.org/~kawasaki/XML-FeedPP-0.43/lib/XML/FeedPP.pm
#
#		NOTE - for Net:Google::AuthSub, if you get a file permissions error from ExtUtil::MakeMaker
#			see this page: http://search.cpan.org/~leont/Module-Build-0.40/lib/Module/Build.pm
#			especially setting the makepl_arg arg for INSTALLMAN1DIR & INSTALLMAN3DIR
#
# SYNTAX:
#       command  : perl ga-api2mpp.pl username\@gmail.com password 1234567
#       cronatab : * */12 * * * perl ga-api2mpp.pl username\@gmail.com password 123467
#
# TO DO:
#       see if I can get filters working, perhaps an array or hash of arrays?
#       see if I can't bring in the profile id(s) from the API
#       use a yaml config file to drive the optional parameters
#		of course I need to not hard-code the dates
#		more robust commandline (getops) handling
#		figure out how to do CDATA blocks with the description & title fields using XML::FeedPP
#		it might be fun to direct output via Template Toolkit;
#		employ utm tracking for outbound links
#			-- such as ?utm_source=ga-app2mpi&utm_medium=rss-widget&utm_campaign=most-popular-pages
#		once you empoy utm tracking, you'll want to employ utm filtering
#		more robust error & exception handling 
#			-- 5 retries on a get_data, duration of each doubles until all fail
#
# LICENSE:
# 		ga-api2mpp.pl by Dean Peters is licensed under a 
#		Creative Commons Attribution-NonCommercial-ShareAlike 3.0 
#		United States License. For more details, read here:
#		http://creativecommons.org/licenses/by-nc-sa/3.0/us/
#
# WARANTEE:
#		absolutely none. use at your own peril
#
# KVETCH:
#		got a comment, complaint, criticism or contribution?
#		http://twitter.com/deanpeters
#		http://healyourchurchwebsite.com
#
##############################################################################
use strict;											# since I'm re-using variables w/in scoped blocks
no warnings 'utf8';									# shaddup, I'm giving a demo dude

# load'up them needed libraries
use Net::Google::Analytics;							# what I use for google analytic calls
use Net::Google::AuthSub;							# I need this to login to ga
use Data::Dumper;									# Keep it around for finding/fixing things
use XML::FeedPP;									# generates the RSS feed

# some local variables we'll use
my $user = $ARGV[0];								# don't forget to escape the \@ in foo\@bar.com
my $pass = $ARGV[1];								# shhh ... secret
my $gaid = $ARGV[2]; 								# this is NOT your GA UA Key
my $domain = 'http://healyourchurchwebsite.com';	# don't let the long name scare ya
my $startdate = '2010-02-23'; 
my $enddate = '2012-02-23';

# instantiate the objects we'll leverage to connect & siphon data
# authetication object
my $auth = Net::Google::AuthSub->new(service => 'analytics');
$auth->login($user, $pass);							# login/connect/authenticate

# analytics object
my $analytics = Net::Google::Analytics->new();
$analytics->auth_params($auth->auth_params);
my $data_feed = $analytics->data_feed;
my $req = $data_feed->new_request();
$req->ids('ga:'.$gaid ); 							# your Analytics profile ID


# here's where we pass arguments for our request
# see: http://code.google.com/apis/analytics/docs/gdata/dimsmets/dimsmets.html
# 
# let's define how we want the rows of data defined
$req->dimensions('ga:pageTitle,ga:pagePath');
# other possible dimensions for future use?
# ga:country			- session
# ga:adwordsCreativeId  - adwords
# ga:searchUsed			- search
# ga:campaign 			- traffic sources

# let's define the columns of data w/in our rows
$req->metrics('ga:uniquePageviews,ga:pageviews,ga:timeOnPage,ga:bounces,ga:entrances,ga:exits');
# other possible metrics for future use?
# ga:bounces			- session
# ga:searchUniques		- search
# ga:uniquePurchases	- ecommerce
# ga:impressions		- adwords
# ga:organicSearches 	- traffic sources

# we don't want EVERYTHING, so let's define ranges & sorting
$req->start_date($startdate);
$req->end_date($enddate);
$req->sort('-ga:uniquePageviews,-ga:pageviews');		# descending order

# Filtering
# http://code.google.com/apis/analytics/docs/gdata/v2/gdataReferenceDataFeed.html#filters
# bah! not working in Net::Google::Analytics -- I knew I shoulda done this in Python!
# $req->max_results('20');
# $req->filters('ga:uniquePageviews>50');

# Go get the feed
my $resp = $data_feed->retrieve($req);

if ( $resp->{"is_success"} eq "1" ) {
 print "Google Analytics API works...OK!\n";
} else {
 print "Google Analytics API has issues...FAIL!\n";
 exit(1);
}

# Do we have enough data?
my $total_results = $resp->{'total_results'};
if ($total_results < 5) {
 print "Google Analytics API works...but failed to find 5 or more records!\n";
 exit(2);
}

# Whoo Hoo! Rock-n-Roll, we gots data, now let's slice-it & dice-it to RSS & CSV files

# set up the channel block of the RSS file
my $feed = XML::FeedPP::RSS->new();
$feed->title( "Most Popular Pages - Heal Your Church Website" );
$feed->link( "http://healyourchurchwebsite.com/test/mpp.rss" );
$feed->pubDate( time() );

# we're going to use this array to help us print a CSV file
my @pages = ("item#\tpageTotalViews\tpageUniqeViews\tpageTimeOnPage\tpageBounces\tpageTitle\tpagePath\n");

# Okay, let's walk the elements 1 row at a time
# ... fyi --> this is where we demo how to progrmatically do fun things with your data
# 
for (my $i = 0; $i < $total_results; $i++) {

	# let's dereference some dimensions (rows) so life is a bit easier
	my $pageTitle = $resp->{'entries'}[$i]->{'dimensions'}[0]->{'value'};
	my $pagePath  = $resp->{'entries'}[$i]->{'dimensions'}[1]->{'value'};
	
	# dereferencing some metrics (columns) -- so lifeis even easier
	my $pageUniqeViews = $resp->{'entries'}[$i]->{'metrics'}[0]->{'value'};
	my $pageTotalViews = $resp->{'entries'}[$i]->{'metrics'}[1]->{'value'};
	my $pageTimeOnPage = $resp->{'entries'}[$i]->{'metrics'}[2]->{'value'};
	my $pageBounces    = $resp->{'entries'}[$i]->{'metrics'}[3]->{'value'};

	# example of post download filtering ...
	# ... especially since we found filtering wasn't working so good w/the perl lib
	next if ($pageTitle =~ m/^\s*$/ || $pagePath =~ m/^\s*$/);	# say no to empty titles & paths
	next if ($pagePath =~ m/^\/$/);								# no need to list the homepage
	next unless $pageUniqeViews > 100;							# let's limit it to really popular pages

	# is it a post page? lets see if there's a date stamp in the path, if so parse it, if not pass on it
	# NOTE *** this assumes your article paths include /yyyy/mm/dd 
	$pagePath =~ m{^/(20[0-1]\d)/(\d{2})/(\d{2})/};
	my ($yy, $mm, $dd) = ($1, $2, $3);							# assign if we find date in path
	next unless ($yy && $mm && $dd);							# no date, no thanks!

	# let's push ROW some data into our CSV file objects
	my $line = "$i\t$pageTotalViews\t$pageUniqeViews\t$pageTimeOnPage\t$pageBounces\t$pageTitle\t$pagePath\n";
	push(@pages, $line);

	# now pushing data into our RSS file object
	if ($i < 20) {											# let's keep the RSS small
		my  $pageDesc = "Item # $i - the article $pageTitle was viewed $pageTotalViews times.";
		my	$item = $feed->add_item( $domain.$pagePath);
			$item->title( $pageTitle );
			$item->description( $pageDesc );
			$item->guid( $domain.$pagePath."#$i" );
			$item->pubDate($yy."-".$mm."-".$dd."T11:30:00+00:00");
	}
}

# write the RSS file
$feed->to_file( "mpp-pl.rss" );

# write the tab delimited file
open(PAGE,">mpp-pl.csv") || die "I can't open mpp.csv"; #open the file to read
print PAGE @pages;
close(PAGE);

# output results to screen
print @pages;
printf ("Total URLs: %s\n", $resp->{'total_results'});

# Yipeee! We're Done -- that's all folks!
exit(0);

